internal/formats/github/encoder.go

```diff
@@ -0,0 +1,154 @@
+package github
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/anchore/packageurl-go"
+	"github.com/anchore/syft/internal"
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/internal/version"
+	"github.com/anchore/syft/syft/pkg"
+	"github.com/anchore/syft/syft/sbom"
+	"github.com/anchore/syft/syft/source"
+)
+
+// toGithubModel converts the provided SBOM to a GitHub dependency model
+func toGithubModel(s *sbom.SBOM) DependencySnapshot {
+	scanTime := time.Now().Format(time.RFC3339) // TODO is there a record of this somewhere?
+	v := version.FromBuild().Version
+	if v == "[not provided]" {
+		v = "0.0.0-dev"
+	}
+	return DependencySnapshot{
+		Version: 0,
+		// TODO allow property input to specify the Job, Sha, and Ref
+		Detector: DetectorMetadata{
+			Name:    internal.ApplicationName,
+			URL:     "https://github.com/anchore/syft",
+			Version: v,
+		},
+		Metadata:  toSnapshotMetadata(s),
+		Manifests: toGithubManifests(s),
+		Scanned:   scanTime,
+	}
+}
+
+// toSnapshotMetadata captures the linux distribution information and other metadata
+func toSnapshotMetadata(s *sbom.SBOM) Metadata {
+	out := Metadata{}
+
+	if s.Artifacts.LinuxDistribution != nil {
+		d := s.Artifacts.LinuxDistribution
+		qualifiers := packageurl.Qualifiers{}
+		if len(d.IDLike) > 0 {
+			qualifiers = append(qualifiers, packageurl.Qualifier{
+				Key:   "like",
+				Value: strings.Join(d.IDLike, ","),
+			})
+		}
+		purl := packageurl.NewPackageURL("generic", "", d.ID, d.VersionID, qualifiers, "")
+		out["syft:distro"] = purl.ToString()
+	}
+
+	return out
+}
+
+// toPath Generates a string representation of the package location, optionally including the layer hash
+func toPath(s source.Metadata, p pkg.Package, full bool) string {
+	if len(p.Locations) > 0 {
+		coords := &p.Locations[0].Coordinates
+		switch s.Scheme {
+		case source.ImageScheme:
+			in := strings.ReplaceAll(s.ImageMetadata.UserInput, ":/", "//")
+			path := strings.TrimPrefix(coords.RealPath, "/")
+			if full {
+				return fmt.Sprintf("%s@%s:/%s", in, coords.FileSystemID, path)
+			}
+			return fmt.Sprintf("%s:/%s", in, path)
+		case source.FileScheme:
+			return fmt.Sprintf("%s:/%s", s.Path, strings.TrimPrefix(coords.RealPath, "/"))
+		case source.DirectoryScheme:
+			return strings.TrimPrefix(fmt.Sprintf("%s/%s", s.Path, coords.RealPath), "./")
+		}
+	}
+	return fmt.Sprintf("%s%s", s.Path, s.ImageMetadata.UserInput)
+}
+
+// toGithubManifests manifests, each of which represents a specific location that has dependencies
+func toGithubManifests(s *sbom.SBOM) Manifests {
+	manifests := map[string]*Manifest{}
+
+	for _, p := range s.Artifacts.PackageCatalog.Sorted() {
+		path := toPath(s.Source, p, false)
+		manifest, ok := manifests[path]
+		if !ok {
+			manifest = &Manifest{
+				Name: path,
+				File: FileInfo{
+					SourceLocation: path,
+				},
+				Metadata: Metadata{
+					"syft:path": toPath(s.Source, p, true),
```

Comment:
The conditionality of `toPath` with the `full` parameter is making this function a bit more difficult for me to understand... I was wondering about what to do about it, but then a question occurred to me — might it makes sense to include the filesystem ID as a separate metadata field here, rather than included it somehow in a single string with the path?

Reply:
agree, originally I was trying to use this as the `file/source_location` but there are currently some issues with that and I don't think that will be viable in the near term